core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,380 @@
+	// [[A, B], [C]] + [[D], [], [E]] = [[A, B, D], [C], [E]]
+	for i := 0; i < len(topics); i++ {
+		if lp.topics[i] == nil {
+			lp.topics[i] = make(map[common.Hash]struct{})
+		}
+		lp.topics[i][topics[i]] = struct{}{}
+	}
+}
+
+func (lp *LogPoller) filterAddresses() []common.Address {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var addresses []common.Address
+	for addr := range lp.addresses {
+		addresses = append(addresses, addr)
+	}
+	sort.Slice(addresses, func(i, j int) bool {
+		return bytes.Compare(addresses[i][:], addresses[j][:]) < 0
+	})
+	return addresses
+}
+
+func (lp *LogPoller) filterTopics() [][]common.Hash {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var topics [][]common.Hash
+	for idx := 0; idx < len(lp.topics); idx++ {
+		var topicPosition []common.Hash
+		for topic := range lp.topics[idx] {
+			topicPosition = append(topicPosition, topic)
+		}
+		sort.Slice(topicPosition, func(i, j int) bool {
+			return bytes.Compare(topicPosition[i][:], topicPosition[j][:]) < 0
+		})
+		topics = append(topics, topicPosition)
+	}
+	return topics
+}
+
+// Replay signals that the poller should resume from a new block.
+func (lp *LogPoller) Replay(fromBlock int64) {
+	lp.replay <- fromBlock
+}
+
+func (lp *LogPoller) Start(parentCtx context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		ctx, cancel := context.WithCancel(parentCtx)
+		lp.ctx = ctx
+		lp.cancel = cancel
+		go lp.run()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		lp.cancel()
+		<-lp.done
+		return nil
+	})
+}
+
+func (lp *LogPoller) Ready() error {
+	return nil
+}
+
+func (lp *LogPoller) Healthy() error {
+	return nil
+}
+
+func (lp *LogPoller) run() {
+	defer close(lp.done)
+	tick := time.After(0)
+	var start int64
+	for {
+		select {
+		case <-lp.ctx.Done():
+			return
+		case fromBlock := <-lp.replay:
+			lp.lggr.Warnw("Replay requested", "from", fromBlock)
+			start = fromBlock
+		case <-tick:
+			tick = time.After(utils.WithJitter(lp.pollPeriod))
+			if start == 0 {
+				lastProcessed, err := lp.orm.SelectLatestBlock(pg.WithParentCtx(lp.ctx))
+				if err != nil {
+					if !errors.Is(err, sql.ErrNoRows) {
+						lp.lggr.Warnw("unable to get starting block", "err", err)
+						continue
+					}
+					// Otherwise this is the first poll _ever_ on a new chain.
+					// Only safe thing to do is to start at finality depth behind tip.
+					latest, err := lp.ec.BlockByNumber(context.Background(), nil)
+					if err != nil {
+						lp.lggr.Warnw("unable to get latest for first poll", "err", err)
+						continue
+					}
+					// Do not support polling chains with don't even have finality depth worth of blocks.
+					// Could conceivably support this but not worth the effort.
+					if int64(latest.NumberU64()) < lp.finalityDepth {
+						lp.lggr.Criticalw("insufficient number of blocks on chain, log poller exiting", "err", err, "latest", latest.NumberU64())
+						return
```

Comment:
Why not just wait till number of blocks exceed the finality depth?
Agreed this is likely never going to be hit, but even if it does, seems like our poller just needs to wait a bit to get finality depth number of blocks without erroring?

Reply:
yeah fair, we could just keep polling